File: trunk/src/yrx.lua
--  
--  (C) 2007 by Remo Dentato (rdentato@users.sourceforge.net)
--
-- Permission to use, copy, modify and distribute this code and
-- its documentation for any purpose is hereby granted without
-- fee, provided that the above copyright notice, or equivalent
-- attribution acknowledgement, appears in all copies and
-- supporting documentation.
-- 
-- Copyright holder makes no representations about the suitability
-- of this software for any purpose. It is provided "as is" without
-- express or implied warranty.
--

local function usage()
  local fmt
  local out
  local use =[[
  usage: yrx [opts] [args]
    Options:
      -i:infile       take input from the specified file 
      -o:outfile      output goes to the specified file
      -a:outfile      output is appended to the specified file
      --              stop processing options
]]
  out={}
  for fmt in pairs(yrx) do 
    if string.sub(fmt,1,1) == "-" then
      out[#out+1] = yrx[fmt](0)
    end
  end
  table.sort(out)
  return use .. "      " .. table.concat(out,'\n      ')

end

--[ [ 
yrx = require("yrxgraph")
yrx = require("yrxasm")
yrx = require("yrxc")
yrx = require("yrxl")
--]]

--yrx=require("yrxlua")

opts = {}
while #arg > 0 do
  opt = nil
  optarg = nil
  string.gsub(arg[1],"^(%-[^:=]+)[:=]?(.*)",
      function (a,b)
        opt=a; optarg = b or ""
      end,1)
  if opt == nil then break  end
  table.remove(arg,1)
  if opt == "--" then break end

  if opt == '-o' or opt == '-a' then
    if opt == '-a' then opt = "a" else opt = "w"  end
    outf = io.open(optarg,opt)
    if outf then 
      io.output(outf)
    else
      err = "YRX ERROR: Unable to open '" .. optarg .. "' for output"
    end
  elseif opt == '-i' then
    inf = io.open(optarg,"r")
    if inf then 
      io.input(inf)
    else
      err = "YRX ERROR: Unable to open '" .. optarg .. "' for input"
    end
  else
    if yrx[opt] then  out = yrx[opt]  end
    opts[opt] = optarg
  end
end

if not err then
  if out then
    err = out(arg,opts)
  else
    err = usage()
  end
end

if err then  io.stderr:write(err) end

if outf then outf:close() end
if inf then inf:close() end

File: trunk/src/yrx.c
/*
**  (C) 2007 by Remo Dentato (rdentato@users.sourceforge.net)
**
** Permission to use, copy, modify and distribute this code and
** its documentation for any purpose is hereby granted without
** fee, provided that the above copyright notice, or equivalent
** attribution acknowledgement, appears in all copies and
** supporting documentation.
**
** Copyright holder makes no representations about the suitability
** of this software for any purpose. It is provided "as is" without
** express or implied warranty.
*/

#include "yrx.h"

void dump_lbl(uint8_t *l)
{
  if (*l == 0) return;

  putchar(*l++);

  while (l[0] <= l[1]) {
    printf("%s",yrxLabelStr(*l++));
    printf("%s",yrxLabelStr(*l++));
  }
}

void dump_tags(uint8_t *t)
{
  if (*t == 0) return;

  printf(" / ");

  while (t[0] != '\0') {
    printf("%s ",yrxTagStr(t[0],t[1]));
    t += 2;
  }
}



void dump(aut *dfa)
{
  uint32_t from;
  Arc *a;

  from = yrxStartState(dfa);

  while (from != 0) {
    while ((a = yrxGetArc(dfa)) != NULL) {
      printf("%5d -> %-5d %p / %p  ", from, a->to, a->lbl, a->tags);
      dump_lbl(yrxArcLabel(a));
      dump_tags(yrxArcTags(a));
      printf("\n");
    }
    from = yrxNextState(dfa);
  }
}

void usage(void)
{
	fprintf(stderr,"Usage: yrxcc rx1 [rx2 ... rx250]\n ");
  exit(1);
}

int main(int argc, char **argv)
{
  aut *dfa;

  if (argc < 2 || 251 < argc) usage();

  dfa = yrxParse(argv+1, argc-1);
  if (dfa != NULL) dump(dfa);

  exit(0);
}

